File: app/src/main/java/com/github/pavlospt/microsoftassignment/AssignmentApp.java
package com.github.pavlospt.microsoftassignment;

import android.app.Application;
import android.content.Context;
import android.support.annotation.NonNull;
import com.github.pavlospt.microsoftassignment.di.components.AssignmentAppComponent;
import com.github.pavlospt.microsoftassignment.di.components.DaggerAssignmentAppComponent;
import com.github.pavlospt.microsoftassignment.di.modules.AssignmentAppModule;
import com.github.pavlospt.microsoftassignment.di.modules.AssignmentStorIOModule;

public class AssignmentApp extends Application {

  private AssignmentAppComponent assignmentAppComponent;

  @Override
  public void onCreate() {
    super.onCreate();
    assignmentAppComponent = DaggerAssignmentAppComponent.builder()
        .assignmentAppModule(new AssignmentAppModule(this))
        .assignmentStorIOModule(new AssignmentStorIOModule())
        .build();
  }

  public AssignmentAppComponent getComponent() {
    return assignmentAppComponent;
  }

  public static AssignmentApp from(@NonNull Context context) {
    return (AssignmentApp) context.getApplicationContext();
  }
}

File: app/src/main/java/com/github/pavlospt/microsoftassignment/AssignmentApp.kt
package com.github.pavlospt.microsoftassignment

import android.app.Application
import android.content.Context
import com.github.pavlospt.microsoftassignment.di.components.AssignmentAppComponent
import com.github.pavlospt.microsoftassignment.di.components.DaggerAssignmentAppComponent
import com.github.pavlospt.microsoftassignment.di.modules.AssignmentAppModule
import com.github.pavlospt.microsoftassignment.di.modules.AssignmentStorIOModule

class AssignmentApp : Application() {

  lateinit var component: AssignmentAppComponent
    private set

  override fun onCreate() {
    super.onCreate()
    component = DaggerAssignmentAppComponent.builder()
        .assignmentAppModule(AssignmentAppModule(this))
        .assignmentStorIOModule(AssignmentStorIOModule())
        .build()
  }

  companion object {

    fun from(context: Context): AssignmentApp {
      return context.applicationContext as AssignmentApp
    }
  }
}
